File: services/research-service/src/main/java/com/researchex/research/service/search/ResearchQueryCriteria.java
package com.researchex.research.service.search;

import lombok.Builder;
import lombok.Getter;
import lombok.ToString;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * 검색 요청을 표현하는 도메인 오브젝트.
 * Controller → Service → Repository 간 명확한 계약을 위해 사용한다.
 */
@Getter
@Builder
@ToString
public class ResearchQueryCriteria {

    /**
     * 사용자가 입력한 전체 텍스트 검색 키워드.
     */
    private final String query;

    /**
     * 멀티 테넌트 환경에서의 테넌트 식별자.
     */
    @Builder.Default
    private final String tenantId = "default";

    /**
     * 진행률 추적을 위한 질의 식별자.
     */
    private final String queryId;

    /**
     * 0-based 페이지 번호.
     */
    private final int page;

    /**
     * 페이지당 문서 수.
     */
    private final int size;

    /**
     * 필터 조건 (field -> values).
     */
    @Builder.Default
    private final Map<String, List<String>> filters = Collections.emptyMap();

    /**
     * 정렬 조건 목록.
     */
    @Builder.Default
    private final List<ResearchSortOption> sorts = Collections.emptyList();

    /**
     * 캐시 사용 여부. 실시간성이 필요한 요청은 false로 내려와서 즉시 Elasticsearch를 조회한다.
     */
    @Builder.Default
    private final boolean useCache = true;

    /**
     * 캐시 키 생성을 위해 page/size/sort/filter 정보를 고유 문자열로 직렬화한다.
     */
    public String toCacheKeySuffix() {
        StringBuilder builder = new StringBuilder();
        builder.append("tenant=").append(Objects.toString(tenantId, "default"))
                .append("|q=").append(Objects.toString(query, ""))
                .append("|p=").append(page)
                .append("|s=").append(size);
        if (!filters.isEmpty()) {
            builder.append("|f=").append(filters.entrySet().stream()
                    .sorted(Map.Entry.comparingByKey())
                    .map(entry -> entry.getKey() + ":" + String.join(",", entry.getValue()))
                    .reduce((a, b) -> a + ";" + b)
                    .orElse(""));
        }
        if (!sorts.isEmpty()) {
            builder.append("|o=").append(sorts.stream()
                    .map(ResearchSortOption::asCacheFragment)
                    .sorted()
                    .reduce((a, b) -> a + "," + b)
                    .orElse(""));
        }
        return builder.toString();
    }
}

File: services/research-service/src/main/kotlin/com/researchex/research/service/search/ResearchQueryCriteria.kt
package com.researchex.research.service.search

/**
 * 검색 요청을 표현하는 도메인 오브젝트.
 * Controller → Service → Repository 간 명확한 계약을 위해 사용한다.
 */
data class ResearchQueryCriteria(
    val query: String?,
    val tenantId: String = "default",
    val queryId: String?,
    val page: Int,
    val size: Int,
    val filters: Map<String, List<String>> = emptyMap(),
    val sorts: List<ResearchSortOption> = emptyList(),
    val useCache: Boolean = true
) {

    /**
     * 캐시 키 생성을 위해 page/size/sort/filter 정보를 고유 문자열로 직렬화한다.
     */
    fun toCacheKeySuffix(): String {
        val builder = StringBuilder()
        builder.append("tenant=").append(tenantId)
            .append("|q=").append(query ?: "")
            .append("|p=").append(page)
            .append("|s=").append(size)

        if (filters.isNotEmpty()) {
            val serializedFilters = filters.entries
                .sortedBy { it.key }
                .joinToString(";") { entry ->
                    "${entry.key}:${entry.value.joinToString(",")}"
                }
            builder.append("|f=").append(serializedFilters)
        }

        if (sorts.isNotEmpty()) {
            val serializedSorts = sorts
                .map { it.asCacheFragment() }
                .sorted()
                .joinToString(",")
            builder.append("|o=").append(serializedSorts)
        }

        return builder.toString()
    }

    companion object {
        fun builder(): Builder = Builder()
    }

    class Builder {
        private var query: String? = null
        private var tenantId: String = "default"
        private var queryId: String? = null
        private var page: Int = 0
        private var size: Int = 20
        private var filters: Map<String, List<String>> = emptyMap()
        private var sorts: List<ResearchSortOption> = emptyList()
        private var useCache: Boolean = true

        fun query(query: String?) = apply { this.query = query }
        fun tenantId(tenantId: String) = apply { this.tenantId = tenantId }
        fun queryId(queryId: String?) = apply { this.queryId = queryId }
        fun page(page: Int) = apply { this.page = page }
        fun size(size: Int) = apply { this.size = size }
        fun filters(filters: Map<String, List<String>>) = apply { this.filters = filters }
        fun sorts(sorts: List<ResearchSortOption>) = apply { this.sorts = sorts }
        fun useCache(useCache: Boolean) = apply { this.useCache = useCache }

        fun build(): ResearchQueryCriteria {
            return ResearchQueryCriteria(
                query = query,
                tenantId = tenantId,
                queryId = queryId,
                page = page,
                size = size,
                filters = filters,
                sorts = sorts,
                useCache = useCache
            )
        }
    }
}
